prow/gcs/reporter/reporter.go

```diff
@@ -0,0 +1,213 @@
+	prowjobErr := gr.reportProwjob(ctx, pj)
+
+	return []*prowv1.ProwJob{pj}, errorutil.NewAggregate(stateErr, prowjobErr)
+}
+
+func (gr *gcsReporter) reportJobState(ctx context.Context, pj *prowv1.ProwJob) error {
+	if pj.Status.State == prowv1.TriggeredState {
+		return gr.reportStartedJob(ctx, pj)
+	}
+	if pj.Complete() {
+		return gr.reportFinishedJob(ctx, pj)
+	}
+	return nil
+}
+
+// reportStartedJob uploads a started.json for the job, iff one did not already exist.
+func (gr *gcsReporter) reportStartedJob(ctx context.Context, pj *prowv1.ProwJob) error {
+	s := metadata.Started{
+		Timestamp: pj.Status.StartTime.Unix(),
+		Metadata:  metadata.Metadata{"uploader": "crier"},
+	}
+	output, err := json.Marshal(s)
+	if err != nil {
+		return fmt.Errorf("failed to marshal started metadata: %v", err)
+	}
+
+	bucketName, dir, err := gr.getJobDestination(pj)
+	if err != nil {
+		return fmt.Errorf("failed to get job destination: %v", err)
+	}
+
+	if gr.dryRun {
+		gr.logger.Infof("Would upload started.json to %q/%q", bucketName, dir)
+		return nil
+	}
+	b := gr.storage.Bucket(bucketName)
+	w := b.Object(path.Join(dir, "started.json")).If(storage.Conditions{DoesNotExist: true}).NewWriter(ctx)
+	return gr.writeContent(w, output)
+}
+
+// reportFinishedJob uploads a finished.json for the job, iff one did not already exist.
+func (gr *gcsReporter) reportFinishedJob(ctx context.Context, pj *prowv1.ProwJob) error {
+	completion := pj.Status.CompletionTime.Unix()
+	passed := pj.Status.State == prowv1.SuccessState
+	f := metadata.Finished{
+		Timestamp: &completion,
+		Passed:    &passed,
+		Metadata:  metadata.Metadata{"uploader": "crier"},
+		Result:    string(pj.Status.State),
+	}
+	output, err := json.Marshal(f)
+	if err != nil {
+		return fmt.Errorf("failed to marshal finished metadata: %v", err)
+	}
+
+	bucketName, dir, err := gr.getJobDestination(pj)
+	if err != nil {
+		return fmt.Errorf("failed to get job destination: %v", err)
+	}
+
+	if gr.dryRun {
+		gr.logger.Infof("Would upload finished.json info to %q/%q", bucketName, dir)
+		return nil
+	}
+	b := gr.storage.Bucket(bucketName)
+	w := b.Object(path.Join(dir, "finished.json")).If(storage.Conditions{DoesNotExist: true}).NewWriter(ctx)
+	return gr.writeContent(w, output)
+}
+
+func (gr *gcsReporter) reportProwjob(ctx context.Context, pj *prowv1.ProwJob) error {
+	// Unconditionally dump the prowjob to GCS, on all job updates.
+	output, err := json.Marshal(pj)
+	if err != nil {
+		return fmt.Errorf("failed to marshal prowjob: %v", err)
+	}
+
+	bucketName, dir, err := gr.getJobDestination(pj)
+	if err != nil {
+		return fmt.Errorf("failed to get job destination: %v", err)
+	}
+
+	if gr.dryRun {
+		gr.logger.Infof("Would upload pod info to %q/%q", bucketName, dir)
+		return nil
+	}
+	b := gr.storage.Bucket(bucketName)
+	w := b.Object(path.Join(dir, "prowjob.json")).NewWriter(ctx)
+	return gr.writeContent(w, output)
+}
+
+func (gr *gcsReporter) writeContent(w *storage.Writer, content []byte) error {
```

Comment:
Seems like this can just be an io.WriterCloser interface rather than a concrete storage.Writer instance (which makes for easier unit testing, although doesn't really seem necessary here)

Reply:
We use some non-WriteCloser properties (`Bucket`, `Name`) for debug reporting, which makes using a WriteCloser annoying enough that I'm inclined not to change it unless we actually write a unit test for this.